File: Android/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/sysinfo/SysInfoFragment.java
package com.didichuxing.doraemonkit.kit.sysinfo;

import android.Manifest;
import android.annotation.TargetApi;
import android.content.pm.PackageInfo;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.View;

import com.blankj.utilcode.util.AppUtils;
import com.blankj.utilcode.util.NetworkUtils;
import com.blankj.utilcode.util.PhoneUtils;
import com.didichuxing.doraemonkit.R;
import com.didichuxing.doraemonkit.ui.base.BaseFragment;
import com.didichuxing.doraemonkit.ui.widget.recyclerview.DividerItemDecoration;
import com.didichuxing.doraemonkit.ui.widget.titlebar.HomeTitleBar;
import com.didichuxing.doraemonkit.util.DeviceUtils;
import com.didichuxing.doraemonkit.util.ExecutorUtil;
import com.didichuxing.doraemonkit.util.PermissionUtil;
import com.didichuxing.doraemonkit.util.UIUtils;

import java.util.ArrayList;
import java.util.List;

/**
 * 手机app信息
 * Created by zhangweida on 2018/6/25.
 */

public class SysInfoFragment extends BaseFragment {
    private RecyclerView mInfoList;
    private SysInfoItemAdapter mInfoItemAdapter;


    @Override
    protected int onRequestLayout() {
        return R.layout.dk_fragment_sys_info;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initView();
        initData();
    }

    private void initView() {
        mInfoList = findViewById(R.id.info_list);
        HomeTitleBar titleBar = findViewById(R.id.title_bar);
        titleBar.setListener(new HomeTitleBar.OnTitleBarClickListener() {
            @Override
            public void onRightClick() {
                getActivity().finish();
            }
        });
        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext());
        mInfoList.setLayoutManager(layoutManager);
        mInfoItemAdapter = new SysInfoItemAdapter(getContext());
        mInfoList.setAdapter(mInfoItemAdapter);
        DividerItemDecoration decoration = new DividerItemDecoration(DividerItemDecoration.VERTICAL);
        decoration.setDrawable(getResources().getDrawable(R.drawable.dk_divider));
        mInfoList.addItemDecoration(decoration);
    }

    private void initData() {
        List<SysInfoItem> sysInfoItems = new ArrayList<>();
        addAppData(sysInfoItems);
        addDeviceData(sysInfoItems);
        if (getContext().getApplicationInfo().targetSdkVersion >= Build.VERSION_CODES.M) {
            addPermissionData(sysInfoItems);
        } else {
            addPermissionDataUnreliable();
        }
        mInfoItemAdapter.setData(sysInfoItems);
    }

    private void addAppData(List<SysInfoItem> sysInfoItems) {
        PackageInfo pi = DeviceUtils.getPackageInfo(getContext());
        sysInfoItems.add(new TitleItem(getString(R.string.dk_sysinfo_app_info)));
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_package_name), pi.packageName));
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_package_version_name), pi.versionName));
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_package_version_code), String.valueOf(pi.versionCode)));
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_package_min_sdk), String.valueOf(getContext().getApplicationInfo().minSdkVersion)));
        }
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_package_target_sdk), String.valueOf(getContext().getApplicationInfo().targetSdkVersion)));
    }

    private void addDeviceData(List<SysInfoItem> sysInfoItems) {
        sysInfoItems.add(new TitleItem(getString(R.string.dk_sysinfo_device_info)));
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_brand_and_model), Build.MANUFACTURER + " " + Build.MODEL));
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_android_version), Build.VERSION.RELEASE + " (" + Build.VERSION.SDK_INT + ")"));
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_ext_storage_free), DeviceUtils.getSDCardSpace(getContext())));
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_rom_free), DeviceUtils.getRomSpace(getContext())));
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_display_size), UIUtils.getWidthPixels() + "x" + UIUtils.getRealHeightPixels()));
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_display_inch), "" + UIUtils.getScreenInch(getActivity())));
        sysInfoItems.add(new SysInfoItem("ROOT", String.valueOf(DeviceUtils.isRoot(getContext()))));
        sysInfoItems.add(new SysInfoItem("DENSITY", String.valueOf(UIUtils.getDensity())));
        sysInfoItems.add(new SysInfoItem("IP", TextUtils.isEmpty(NetworkUtils.getIPAddress(true)) ? "null" : NetworkUtils.getIPAddress(true)));
        sysInfoItems.add(new SysInfoItem("Mac", TextUtils.isEmpty(com.blankj.utilcode.util.DeviceUtils.getMacAddress()) ? "null" : com.blankj.utilcode.util.DeviceUtils.getMacAddress()));
        try {
            sysInfoItems.add(new SysInfoItem("IMEI", TextUtils.isEmpty(PhoneUtils.getIMEI()) ? "null" : PhoneUtils.getIMEI()));
        } catch (Exception e) {
        }
        sysInfoItems.add(new SysInfoItem("Sign MD5", AppUtils.getAppSignatureMD5()));
        sysInfoItems.add(new SysInfoItem("Sign SHA1", AppUtils.getAppSignatureSHA1()));
        sysInfoItems.add(new SysInfoItem("Sign SHA256", AppUtils.getAppSignatureSHA256()));


    }

    /**
     * 不可靠的检测权限方式
     */
    private void addPermissionDataUnreliable() {
        ExecutorUtil.execute(new Runnable() {
            @Override
            public void run() {
                final List<SysInfoItem> list = new ArrayList<>();
                list.add(new TitleItem(getString(R.string.dk_sysinfo_permission_info_unreliable)));
                list.add(new SysInfoItem(getString(R.string.dk_sysinfo_permission_location), PermissionUtil.checkLocationUnreliable(getContext()) ? "YES" : "NO", true));
                list.add(new SysInfoItem(getString(R.string.dk_sysinfo_permission_sdcard), PermissionUtil.checkStorageUnreliable() ? "YES" : "NO", true));
                list.add(new SysInfoItem(getString(R.string.dk_sysinfo_permission_camera), PermissionUtil.checkCameraUnreliable() ? "YES" : "NO", true));
                list.add(new SysInfoItem(getString(R.string.dk_sysinfo_permission_record), PermissionUtil.checkRecordUnreliable() ? "YES" : "NO", true));
                list.add(new SysInfoItem(getString(R.string.dk_sysinfo_permission_read_phone), PermissionUtil.checkReadPhoneUnreliable(getContext()) ? "YES" : "NO", true));
                list.add(new SysInfoItem(getString(R.string.dk_sysinfo_permission_contact), PermissionUtil.checkReadContactUnreliable(getContext()) ? "YES" : "NO", true));
                getView().post(new Runnable() {
                    @Override
                    public void run() {
                        if (SysInfoFragment.this.isDetached()) {
                            return;
                        }
                        mInfoItemAdapter.append(list);
                    }
                });
            }
        });
    }

    @TargetApi(Build.VERSION_CODES.M)
    private void addPermissionData(List<SysInfoItem> sysInfoItems) {
        sysInfoItems.add(new TitleItem(getString(R.string.dk_sysinfo_permission_info)));
        String[] p1 = {
                Manifest.permission.ACCESS_COARSE_LOCATION,
                Manifest.permission.ACCESS_FINE_LOCATION
        };
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_permission_location), checkPermission(p1), true));
        String[] p2 = {
                Manifest.permission.READ_EXTERNAL_STORAGE,
                Manifest.permission.WRITE_EXTERNAL_STORAGE
        };
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_permission_sdcard), checkPermission(p2), true));
        String[] p3 = {
                Manifest.permission.CAMERA
        };
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_permission_camera), checkPermission(p3), true));
        String[] p4 = {
                Manifest.permission.RECORD_AUDIO
        };
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_permission_record), checkPermission(p4), true));
        String[] p5 = {
                Manifest.permission.READ_PHONE_STATE
        };
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_permission_read_phone), checkPermission(p5), true));
        String[] p6 = {
                Manifest.permission.READ_CONTACTS
        };
        sysInfoItems.add(new SysInfoItem(getString(R.string.dk_sysinfo_permission_contact), checkPermission(p6), true));
    }

    private String checkPermission(String... perms) {
        return PermissionUtil.hasPermissions(getContext(), perms) ? "YES" : "NO";
    }

}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/sysinfo/SysInfoFragment.kt
package com.didichuxing.doraemonkit.kit.sysinfo

import android.Manifest
import android.annotation.SuppressLint
import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.blankj.utilcode.util.AppUtils
import com.blankj.utilcode.util.NetworkUtils
import com.blankj.utilcode.util.PhoneUtils
import com.didichuxing.doraemonkit.R
import com.didichuxing.doraemonkit.extensions.runBelowAndroidM
import com.didichuxing.doraemonkit.extensions.runGreaterThanAndroidM
import com.didichuxing.doraemonkit.extensions.runGreaterThanAndroidN
import com.didichuxing.doraemonkit.kit.core.BaseFragment
import com.didichuxing.doraemonkit.util.DeviceUtils
import com.didichuxing.doraemonkit.util.DeviceUtils.getRomSpace
import com.didichuxing.doraemonkit.util.DeviceUtils.getSDCardSpace
import com.didichuxing.doraemonkit.util.ExecutorUtil
import com.didichuxing.doraemonkit.util.PermissionUtil.checkCameraUnreliable
import com.didichuxing.doraemonkit.util.PermissionUtil.checkLocationUnreliable
import com.didichuxing.doraemonkit.util.PermissionUtil.checkReadContactUnreliable
import com.didichuxing.doraemonkit.util.PermissionUtil.checkReadPhoneUnreliable
import com.didichuxing.doraemonkit.util.PermissionUtil.checkRecordUnreliable
import com.didichuxing.doraemonkit.util.PermissionUtil.checkStorageUnreliable
import com.didichuxing.doraemonkit.util.PermissionUtil.hasPermissions
import com.didichuxing.doraemonkit.util.UIUtils.density
import com.didichuxing.doraemonkit.util.UIUtils.getScreenInch
import com.didichuxing.doraemonkit.util.UIUtils.realHeightPixels
import com.didichuxing.doraemonkit.util.UIUtils.widthPixels
import com.didichuxing.doraemonkit.widget.recyclerview.DividerItemDecoration
import com.didichuxing.doraemonkit.widget.titlebar.HomeTitleBar
import java.util.ArrayList

/**
 * @author lostjobs created on 2020/6/9
 */
class SysInfoFragment : BaseFragment(), HomeTitleBar.OnTitleBarClickListener {

  private var mInfoList: RecyclerView? = null
  private val mInfoItemAdapter by lazy {
    SysInfoItemAdapter()
  }

  override fun onRequestLayout(): Int = R.layout.dk_fragment_sys_info

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    kotlin.runCatching {
      initView()
      initData()
    }
  }

  private fun initView() {
    mInfoList = view?.findViewById(R.id.info_list)
    mInfoList?.apply {
      layoutManager = LinearLayoutManager(requireContext())
      adapter = mInfoItemAdapter
      addItemDecoration(
        DividerItemDecoration(DividerItemDecoration.VERTICAL).setDrawable(
          ContextCompat.getDrawable(requireContext(), R.drawable.dk_divider)
        )
      )
    }

    findViewById<HomeTitleBar>(R.id.title_bar).mListener = this
  }

  private fun initData() {
    val sysInfoItems = mutableListOf<SysInfoItem>()
    addAppData(sysInfoItems)
    addDeviceData(sysInfoItems)
    runGreaterThanAndroidM {
      addPermissionData(sysInfoItems)
    }
    runBelowAndroidM {
      addPermissionDataUnreliable()
    }
    mInfoItemAdapter.setData(sysInfoItems)
  }

  private fun addPermissionDataUnreliable() {
    ExecutorUtil.execute {
      val list: MutableList<SysInfoItem> = ArrayList()
      list.add(SysInfoItem(title = getString(R.string.dk_sysinfo_permission_info_unreliable)))
      list.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = getString(R.string.dk_sysinfo_permission_location),
          value = if (checkLocationUnreliable(context!!)) "YES" else "NO",
          isPermission = true
        )
      )
      list.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = getString(R.string.dk_sysinfo_permission_sdcard),
          value = if (checkStorageUnreliable()) "YES" else "NO",
          isPermission = true
        )
      )
      list.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = getString(R.string.dk_sysinfo_permission_camera),
          value = if (checkCameraUnreliable()) "YES" else "NO",
          isPermission = true
        )
      )
      list.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = getString(R.string.dk_sysinfo_permission_record),
          value = if (checkRecordUnreliable()) "YES" else "NO",
          isPermission = true
        )
      )
      list.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = getString(R.string.dk_sysinfo_permission_read_phone),
          value = if (checkReadPhoneUnreliable(context!!)) "YES" else "NO",
          isPermission = true
        )
      )
      list.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = getString(R.string.dk_sysinfo_permission_contact),
          value = if (checkReadContactUnreliable(context!!)) "YES" else "NO",
          isPermission = true
        )
      )
      view?.post {
        if (!this@SysInfoFragment.isDetached) {
          mInfoItemAdapter.append(list)
        }
      }
    }
  }

  private fun addPermissionData(sysInfoItems: MutableList<SysInfoItem>) {
    sysInfoItems.add(SysInfoItem(title = getString(R.string.dk_sysinfo_permission_info)))
    val p1 = arrayOf(
      Manifest.permission.ACCESS_COARSE_LOCATION,
      Manifest.permission.ACCESS_FINE_LOCATION
    )
    sysInfoItems.add(
      SysInfoItem(
        type = SysInfoItemAdapter.TYPE_INFO_ITEM,
        title = getString(R.string.dk_sysinfo_permission_location),
        value = checkPermission(*p1),
        isPermission = true
      )
    )
    val p2 = arrayOf(
      Manifest.permission.READ_EXTERNAL_STORAGE,
      Manifest.permission.WRITE_EXTERNAL_STORAGE
    )
    sysInfoItems.add(
      SysInfoItem(
        type = SysInfoItemAdapter.TYPE_INFO_ITEM,
        title = getString(R.string.dk_sysinfo_permission_sdcard),
        value = checkPermission(*p2),
        isPermission = true
      )
    )
    val p3 = arrayOf(
      Manifest.permission.CAMERA
    )
    sysInfoItems.add(
      SysInfoItem(
        type = SysInfoItemAdapter.TYPE_INFO_ITEM,
        title = getString(R.string.dk_sysinfo_permission_camera),
        value = checkPermission(*p3),
        isPermission = true
      )
    )
    val p4 = arrayOf(
      Manifest.permission.RECORD_AUDIO
    )
    sysInfoItems.add(
      SysInfoItem(
        type = SysInfoItemAdapter.TYPE_INFO_ITEM,
        title = getString(R.string.dk_sysinfo_permission_record),
        value = checkPermission(*p4),
        isPermission = true
      )
    )
    val p5 = arrayOf(
      Manifest.permission.READ_PHONE_STATE
    )
    sysInfoItems.add(
      SysInfoItem(
        type = SysInfoItemAdapter.TYPE_INFO_ITEM,
        title = getString(R.string.dk_sysinfo_permission_read_phone),
        value = checkPermission(*p5),
        isPermission = true
      )
    )
    val p6 = arrayOf(
      Manifest.permission.READ_CONTACTS
    )
    sysInfoItems.add(
      SysInfoItem(
        type = SysInfoItemAdapter.TYPE_INFO_ITEM,
        title = getString(R.string.dk_sysinfo_permission_contact),
        value = checkPermission(*p6),
        isPermission = true
      )
    )
  }

  @SuppressLint("MissingPermission")
  private fun addDeviceData(sysInfoItems: MutableList<SysInfoItem>) {
    sysInfoItems.add(SysInfoItem(title = getString(R.string.dk_sysinfo_device_info)))
    sysInfoItems.add(
      SysInfoItem(
        type = SysInfoItemAdapter.TYPE_INFO_ITEM,
        title = getString(R.string.dk_sysinfo_brand_and_model),
        value = Build.MANUFACTURER + " " + Build.MODEL
      )
    )
    sysInfoItems.add(
      SysInfoItem(
        type = SysInfoItemAdapter.TYPE_INFO_ITEM,
        title = getString(R.string.dk_sysinfo_android_version),
        value = Build.VERSION.RELEASE + " (" + Build.VERSION.SDK_INT + ")"
      )
    )
    kotlin.runCatching {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = getString(R.string.dk_sysinfo_ext_storage_free),
          value = getSDCardSpace(context!!)
        )
      )
    }
    kotlin.runCatching {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = getString(R.string.dk_sysinfo_rom_free),
          value = getRomSpace(context!!)
        )
      )
    }
    kotlin.runCatching {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = getString(R.string.dk_sysinfo_display_size),
          value = widthPixels.toString() + "x" + realHeightPixels
        )
      )
    }
    kotlin.runCatching {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = getString(R.string.dk_sysinfo_display_inch),
          value = "" + getScreenInch(activity!!)
        )
      )
    }
    kotlin.runCatching {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = "ROOT",
          value = com.blankj.utilcode.util.DeviceUtils.isDeviceRooted().toString()
        )
      )
    }
    kotlin.runCatching {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = "DENSITY",
          value = density.toString()
        )
      )
    }
    kotlin.runCatching {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = "IP",
          value = if (TextUtils.isEmpty(NetworkUtils.getIPAddress(true))) "null" else NetworkUtils.getIPAddress(
            true
          )
        )
      )
    }
    kotlin.runCatching {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = "Mac",
          value = if (TextUtils.isEmpty(com.blankj.utilcode.util.DeviceUtils.getMacAddress())) "null" else com.blankj.utilcode.util.DeviceUtils.getMacAddress()
        )
      )
    }
    kotlin.runCatching {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = "IMEI",
          value = if (TextUtils.isEmpty(PhoneUtils.getIMEI())) "null" else PhoneUtils.getIMEI()
        )
      )
    }
    kotlin.runCatching {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = "Sign MD5",
          value = AppUtils.getAppSignatureMD5()
        )
      )
    }
    kotlin.runCatching {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = "Sign SHA1",
          value = AppUtils.getAppSignatureSHA1()
        )
      )
    }
    kotlin.runCatching {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = "Sign SHA256",
          value = AppUtils.getAppSignatureSHA256()
        )
      )
    }
  }

  @SuppressLint("NewApi")
  private fun addAppData(sysInfoItems: MutableList<SysInfoItem>) {
    val pi = DeviceUtils.getPackageInfo(requireContext())
    sysInfoItems.add(SysInfoItem(title = getString(R.string.dk_sysinfo_app_info)))
    sysInfoItems.add(
      SysInfoItem(
        type = SysInfoItemAdapter.TYPE_INFO_ITEM,
        title = getString(R.string.dk_sysinfo_package_name),
        value = pi?.packageName
      )
    )
    sysInfoItems.add(
      SysInfoItem(
        type = SysInfoItemAdapter.TYPE_INFO_ITEM,
        title = getString(R.string.dk_sysinfo_package_version_name),
        value = pi?.versionName
      )
    )
    sysInfoItems.add(
      SysInfoItem(
        type = SysInfoItemAdapter.TYPE_INFO_ITEM,
        title = getString(R.string.dk_sysinfo_package_version_code),
        value = pi?.versionCode.toString()
      )
    )
    runGreaterThanAndroidN {
      sysInfoItems.add(
        SysInfoItem(
          type = SysInfoItemAdapter.TYPE_INFO_ITEM,
          title = getString(R.string.dk_sysinfo_package_min_sdk),
          value = requireContext().applicationInfo.minSdkVersion.toString()
        )
      )
    }
    sysInfoItems.add(
      SysInfoItem(
        type = SysInfoItemAdapter.TYPE_INFO_ITEM,
        title = getString(R.string.dk_sysinfo_package_target_sdk),
        value = requireContext().applicationInfo.targetSdkVersion.toString()
      )
    )
  }

   private fun checkPermission(vararg perms: String): String? {
    try {
      return if (hasPermissions(requireContext(), *perms)) "YES" else "NO"
    } catch (e: NullPointerException) {
      e.printStackTrace()
    }
    return "NO"
  }

  override fun onRightClick() {
    requireActivity().finish()
  }
}